### PR TITLE
Eden to evetest: rewrite networking, security, apps, lps and storage tests

### PR DESCRIPTION
**Pull request by milan-zededa**
# Description

Rewrites a batch of [Eden](https://github.com/lf-edge/eden) E2E test scenarios (networking, security, apps, LPS, upgrade, storage)
as native Go tests in the `evetest` framework, plus some framework/library enhancements
and one SDN bugfix needed to support them.

## Commits

- **`zedrouter, zedagent: rename NI status field to CombinedNTPServers`**
  `NetworkInstanceConfig.NtpServers` (the NI's own raw configured list) and
  `NetworkInstanceStatus.NTPServers` (the combined NI+port list zedrouter
  actually maintains) differed only by case, so zedagent's reporting code
  silently picked up the wrong one via field promotion through the embedded
  config struct. Renamed the status field to `CombinedNTPServers` and fixed
  zedagent to publish it.

- **`evetest/sdn: fix dnsmasq router option not being suppressed`**
  A DHCP server config with `GatewayIPv4` left nil (the documented way to
  make clients skip installing a default IPv4 route) simply omitted the
  `dhcp-option=option:router` line instead of explicitly suppressing it.
  dnsmasq's own default, absent an explicit override, is to advertise its
  own listening address on that subnet as the router, so clients still
  received one anyway.

- **`evetest: framework/library enhancements`**
  Adds the `EdgeDevice`/`EdgeCluster`/harness capabilities the test
  rewrites below need: flow-log streaming (`IterateDeviceFlowLogs`,
 `SubscribeToDeviceFlowLogs`, the gRPC server iterator, and the
  `evetest eve flow-logs` CLI subcommand), `EdgeDevice`: `GetArch`,
 `PowerOff`, `PowerOn`, `WaitForClusterNodeIsReady`,
  and `EdgeDevice.DialViaSSH` (an SSH-tunneled `net.Conn`, used to reach
  the Kubevirt VNC proxy which only binds to the device's loopback).
  Also removes the now-unused `SkipIfHypervisorKubevirt` helper (all tests
  now support running at-least single-node EVE-K).

- **`evetest: rewrite eden test suites (networking, security, apps, lps, upgrade)`**
  The actual eden -> evetest ports:
  - **networking**: many new tests, now clearly surpassing eden in the coverage
    of EVE networking.
  - **security**: `TestAppArmorEnabled` (kernel AppArmor status) and
    `TestVCom` (vcomlink TPM-over-vsock request/response from inside a
    guest VM).
  - **apps**: `TestVNC` (console access, VM and container apps) and app
    purge.
  - **LPS**: local profile server scenarios (app/device local info,
    profile switching, radio silence).
  - **upgrade**: support upgrade scenario with OCI datastore.

- **`evetest: port eden storage tests, add new datastore coverage`**
    Rewrite eden's storage/volume test suites as evetest tests under
    evetest/tests/storage/:
      - tests/eclient/testdata/mount.txt -> TestMountedVolumes
      - tests/eclient/testdata/disk.txt -> TestExtraDiskAttach
      - tests/volume/testdata/volumes_test.txt -> TestVolumes
      - tests/zfs/testdata/state_and_layout_check.txt -> TestZFSDiskLayout
      - tests/volume/testdata/volume_sftp.txt -> merged into
        TestSFTPDatastore (tests/networking/datastore_test.go), since it
        overlapped with the new datastore tests below

- **evetest: run an embedded OCI registry for container-image tests**
    Both BaseOSDatastoreOCI upgrades and TestContainerRegistry previously
    required their target image to already be published to a real, externally
    reachable registry (Docker Hub), so they broke for an unpublished local/dev
    build and depended on the device having Internet access.

- **`evetest: sweep stale img-cache-* dirs left by killed containers`**

This leaves only the "smoke" test suite as not yet ported from eden to evetest.

## How to test and validate this PR

```sh
# networking
make evetest NAME=TestFlowLog
make evetest NAME=TestNetworkConfigFallback
make evetest NAME=TestPortFailover
make evetest NAME=TestIntermittentConnectivity
make evetest NAME=TestMgmtTrafficRoutedViaApp
make evetest NAME=TestNIReplace
make evetest NAME=TestMoveAppBetweenNIs
make evetest NAME=TestAirGapSwitchNI
make evetest NAME=TestLimitedIPSpace
make evetest NAME=TestDeviceNTPConfig
make evetest NAME=TestApplicationNTPConfig

# security
make evetest NAME=TestAppArmorEnabled
make evetest NAME=TestVCom

# apps
make evetest NAME=TestVNC
make evetest NAME=TestPurgeNeverActivatedApp

# lps (local profile server)
make evetest NAME=TestAppLocalInfo
make evetest NAME=TestDevLocalInfo
make evetest NAME=TestProfile
make evetest NAME=TestRadioSilence

# upgrade
EVETEST_DATASTORE_TYPE=OCI make evetest NAME=TestEVEUpgrade

# storage
make evetest NAME=TestVolumes
make evetest NAME=TestMountedVolumes
make evetest NAME=TestExtraDiskAttach
make evetest NAME=TestZFSDiskLayout

# datastore
make evetest NAME=TestHTTPDatastore
make evetest NAME=TestHTTPSDatastore
make evetest NAME=TestSFTPDatastore
make evetest NAME=TestContainerRegistry
```

## Changelog notes

Fixed a bug where a network instance's device-level NTP servers (learned
from DHCP or configured directly on the port) were not included in the
NTP server list reported to the controller for apps attached to that
network instance -- only the network instance's own explicitly
configured NTP servers were. Apps now see the full, correct combined
list.

All other changes are test-only (evetest framework additions and eden
test-suite rewrites) plus one SDN test-harness bugfix; no other
user-facing EVE behavior changes.

## PR Backports

```text
- 17.0-stable: Eventually to be backported (the entire evetest dir).
- 16.0-stable: No, as evetest is not available there.
- 14.5-stable: No, as evetest is not available there.
- 13.4-stable: No, as evetest is not available there.
```

## Checklist

- [x] I've provided a proper description
- [ ] I've added the proper documentation
- [x] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [x] I've written the test verification instructions
- [x] I've set the proper labels to this PR
- [x] I've checked the boxes above, or I've provided a good reason why I didn't check them.

### CODE REVIEW

**Comment by milan-zededa** [comment]
@claude

File: evetest/ssh.go [hunk 1]
```diff
@@ -54,6 +54,64 @@
...
+	signer, err := ssh.ParsePrivateKey(keyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse client key: %w", err)
+	}
+
+	addr := net.JoinHostPort(eveIP, "22")
+	sshConfig := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply evetest/ssh.go:86]: This is OK, these are ephemeral EVE VMs used only for testing. So we use the same SSH key in all tests.

File: evetest/edgedevice.go [hunk 2]
```diff
@@ -407,43 +477,60 @@ func (d *EdgeDevice) UpgradeEVE(targetEVEVersion string, targetEVEHypervisor Hyp
...
-			d.th.t.Fatalf("Failed to extract EVE rootfs from %s: %v",
-				imageName, err)
+	// Build upgrade device config from a clone of the current config.
+	config := d.GetConfig()
+	switch datastoreType {
+	case BaseOSDatastoreOCI:
+		// Let EVE pull the rootfs directly from the same OCI registry the
+		// target image was tagged in -- no local extraction or HTTP hosting
+		// needed. imageName is "<repo>:<tag>" (see utils.EVEDockerImageName).
+		repo, tag, found := strings.Cut(imageName, ":")
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply evetest/edgedevice.go]: Fixed

**Comment by milan-zededa** [comment]
@claude

**Review by eriknordmark — CHANGES_REQUESTED** [review]
Yetus reports 
|  -1  |    golangcilint  |   8m 17s   | The patch generated 24 new + 538 
|      |                  |            | unchanged - 0 fixed = 562 total (was
|      |                  |            | 538)
but the summary and annotations appear to be showing unchanged files with yetus issues. @rene I see this with yetus on other PRs as well - as soon as there is a -1 vote we end up with a summary and annotations which do not relate to the diffs in the PR.

Needs to be rebased and conflicts resolved - hence "changes requested".

Also, how soon can we get the running of these tests into the EVE CI? And can we make that be included in the code coverage so we don't show zero coverage as we do for this PR?

**Comment by milan-zededa** [comment]
@eriknordmark @rene 
I have rebased the PR to fix merge conflicts, added storage tests rewritten from eden to evetest, and addressed all Yetus issues relevant to this PR.
The PR is therefore ready for merge.

> Also, how soon can we get the running of these tests into the EVE CI? And can we make that be included in the code coverage so we don't show zero coverage as we do for this PR?

We do not have enough hardware resources to run both eden and evetest. So we need to first complete the migration of tests from eden to evetest. The only remaining (after this PR) are smoke tests and one storage test, which Rene, Andrew and Paul are taking care of. Once that has completed, we can migrate to evetest.
We could also enable code coverage collection (it is already supported in evetest), but it will certainly have some effect on performance and test duration. Perhaps it makes sense to have this enabled only on some "large" workflow that would run all tests once per night for the master branch, as opposed to collecting it for every PR separately.

File: evetest/localregistry.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+// newLocalRegistryHandler returns an http.Handler implementing the Docker
+// Registry HTTP API v2 (in-memory blob/manifest store), mounted at "/v2/" on
+// the harness's image-server listeners alongside the plain file server. It
+// lets EVE pull a container image (an upgrade rootfs, or an application
+// content tree) directly from evetest, without that image ever having been
+// published to a real, reachable registry -- see
+// PushDockerImageToLocalRegistry.
+func newLocalRegistryHandler() http.Handler {
+	return registry.New()
```
Comment (europaul) [inline evetest/localregistry.go]: `registry.New()` in v0.14.0 hardcodes the in-memory blob handler, so the pushed image (an `lfedge/eve` image for the OCI upgrade variant -- hundreds of MB) stays in the harness heap for the whole process. `TestUpgradeSuite` runs all variants under a single `Init`, so it is still resident during the later ones.

Bumping go-containerregistry to >= v0.19.1 gets you `registry.WithBlobHandler(registry.NewDiskBlobHandler(dir))`, which would put the blobs under `th.imgServerDir` instead. Probably worth doing anyway now that the module is a direct dependency -- v0.14.0 is from early 2023.

File: evetest/localregistry.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+// newLocalRegistryHandler returns an http.Handler implementing the Docker
+// Registry HTTP API v2 (in-memory blob/manifest store), mounted at "/v2/" on
+// the harness's image-server listeners alongside the plain file server. It
+// lets EVE pull a container image (an upgrade rootfs, or an application
+// content tree) directly from evetest, without that image ever having been
+// published to a real, reachable registry -- see
+// PushDockerImageToLocalRegistry.
+func newLocalRegistryHandler() http.Handler {
```
Comment (europaul) [inline evetest/localregistry.go]: `registry.New()` defaults all of its loggers to `log.New(os.Stderr, ...)`, so every manifest/blob request is printed outside logrus and outside the artifact dir, interleaved with `go test` output.

`registry.Logger()` exists in v0.14.0 -- passing one backed by the harness logger fixes it (needs a logger argument here).

File: evetest/localregistry.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+// via BaseOSDatastoreOCI, or a DockerContainer volume/app image) without the
+// image under test already being published to a real, externally reachable
+// registry: evetest re-serves whatever the local Docker daemon has under
+// imageName ("<repo>:<tag>", e.g. as returned by utils.EVEDockerImageName)
+// as a datastore of its own.
+func PushDockerImageToLocalRegistry(imageName string) (DockerContainer, error) {
+	th := getTestHarness()
+	log := th.log.WithField("component", "local-registry")
+
+	repo, tag, ok := strings.Cut(imageName, ":")
```
Comment (europaul) [inline evetest/localregistry.go]: `strings.Cut` splits at the *first* colon, whereas the code this replaces in `UpgradeEVE` used `strings.LastIndex` deliberately. A repo carrying a registry port (`harbor.example.com:5000/lfedge/eve:1.2.3-kvm-amd64`) now yields `repo="harbor.example.com"`.

`LastIndex` alone would not be enough either -- the re-tag below would then produce `10.x.x.x:80/harbor.example.com:5000/...`. Parsing with `name.ParseReference` and re-tagging from `.Context().RepositoryStr()` handles both, and `name` is already imported.

**Review by eriknordmark — APPROVED** [review]
Should we do something about the yetus issues against evetest? They might not have been introduced by this PR (and we have an issue with the summary/annotations being all over the place), but makes sense to take a look at the summary in https://github.com/lf-edge/eve/actions/runs/31085972149?pr=6244

Kick off regression tests for good measure.

File: evetest/localregistry.go (hunk 3 above)
Comment (milan-zededa) [reply evetest/localregistry.go]: Done

File: evetest/localregistry.go (hunk 4 above)
Comment (milan-zededa) [reply evetest/localregistry.go]: Done

File: evetest/localregistry.go (hunk 5 above)
Comment (milan-zededa) [reply evetest/localregistry.go]: Fixed

File: evetest/localregistry.go (hunk 4 above)
Comment (milan-zededa) [reply evetest/localregistry.go]: Now it logs like this:
```
HARNESS  time="2026-08-06T12:55:18Z" level=debug msg="OCI Registry: GET /v2/lfedge/evetest-ubuntu-ctr/blobs/sha256:bfb639d8897b6a35afa63552a945cf1e821a2bd4722ae9c3f0d01f19e8e62664"
HARNESS  time="2026-08-06T12:55:18Z" level=debug msg="OCI Registry: GET /v2/"
HARNESS  time="2026-08-06T12:55:18Z" level=debug msg="OCI Registry: GET /v2/lfedge/evetest-ubuntu-ctr/blobs/sha256:ad8d8f2cc1c08a1b0f5ead34e57d34e28e7b9eb51e2a82c1cb8f3ef7bfb2d6a5"
HARNESS  time="2026-08-06T12:55:18Z" level=debug msg="OCI Registry: HEAD /v2/lfedge/evetest-ubuntu-ctr/blobs/sha256:bfb639d8897b6a35afa63552a945cf1e821a2bd4722ae9c3f0d01f19e8e62664"
HARNESS  time="2026-08-06T12:55:18Z" level=debug msg="OCI Registry: HEAD /v2/lfedge/evetest-ubuntu-ctr/blobs/sha256:1d2c5bc98e8170a8e5ce3776edba944097c26bc35fe7c52aa2a2ed905746325d"
```

**Comment by milan-zededa** [comment]
> Should we do something about the yetus issues against evetest? They might not have been introduced by this PR (and we have an issue with the summary/annotations being all over the place), but makes sense to take a look at the summary in https://github.com/lf-edge/eve/actions/runs/31085972149?pr=6244
> 
> Kick off regression tests for good measure.

I will address them at some point, but in a separate PR (there are 140 issues raised for evetest, mostly `Error return value of <something> is not checked`)

**Comment by rene** [comment]
The yetus issue is related to an error with bufcompat plugin:

`Failure: input contained 2 images, whereas against contained 1 images`

We use `--buf-basedir=/workspace/src/evetest` but inside evetest we have more than one protbuf module:

- evetest/grpcapi/eve-api
- evetest/grpcapi/proto

That's why 2 images are generated, the plugin only supports check against one... I will open a PR with the fix, we should point only to `evetest/grpcapi/proto`

**Comment by milan-zededa** [comment]
> The yetus issue is related to an error with bufcompat plugin:
> 
> `Failure: input contained 2 images, whereas against contained 1 images`
> 
> We use `--buf-basedir=/workspace/src/evetest` but inside evetest we have more than one protbuf module:
> 
>     * evetest/grpcapi/eve-api
> 
>     * evetest/grpcapi/proto
> 
> 
> That's why 2 images are generated, the plugin only supports check against one... I will open a PR with the fix, we should point only to `evetest/grpcapi/proto`

I tried to fix this some time ago, but could never avoid the -1 for bufcompat. Maybe your agent will do better job :)

**Comment by rene** [comment]
Yetus fix: https://github.com/lf-edge/eve/pull/6276

**Comment by rene** [comment]
@milan-zededa  Yetus fix is merged. You can rebase if you want.

**Comment by milan-zededa** [comment]
> @milan-zededa Yetus fix is merged. You can rebase if you want.

@rene Rebased

**Comment by rene** [comment]
Yetus passed. ~There is no reason to wait for Eden tests or builds since changes are only on evetest.~

My bad, I forgot the small changes on pillar as well and just merged. @milan-zededa will keep an eye on the tests.
